### PR TITLE
fix: Correctly handle custom deps and task_group during DAG serialization #16690

### PR DESCRIPTION
**Pull request by luoyuliuyin**
We check if the dag changed or not via dag_hash, so we need to correctly handle deps and task_group during DAG serialization to ensure that the generation of dag_hash is stable.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/serialization/serialized_objects.py [hunk 1]
```diff
@@ -808,10 +808,10 @@ def serialize_task_group(cls, task_group: TaskGroup) -> Optional[Union[Dict[str,
                 else (DAT.TASK_GROUP, SerializedTaskGroup.serialize_task_group(child))
                 for label, child in task_group.children.items()
             },
-            "upstream_group_ids": cls._serialize(list(task_group.upstream_group_ids)),
-            "downstream_group_ids": cls._serialize(list(task_group.downstream_group_ids)),
-            "upstream_task_ids": cls._serialize(list(task_group.upstream_task_ids)),
-            "downstream_task_ids": cls._serialize(list(task_group.downstream_task_ids)),
+            "upstream_group_ids": cls._serialize(sorted(list(task_group.upstream_group_ids))),
```
Comment (uranusjr) [inline airflow/serialization/serialized_objects.py]:
```suggestion
            "upstream_group_ids": cls._serialize(sorted(task_group.upstream_group_ids)),
```

Same for others below.

File: tests/serialization/test_dag_serialization.py [hunk 2]
```diff
@@ -978,6 +978,107 @@ def check_task_group(node):
...
+            task1 >> task2
+
+        serialize_op = SerializedBaseOperator.serialize_operator(dag.task_dict["task1"])
+        deps = serialize_op["deps"]
+        assert len(deps) == 5
+        assert deps[0] == 'airflow.ti_deps.deps.not_in_retry_period_dep.NotInRetryPeriodDep'
+        assert deps[1] == 'airflow.ti_deps.deps.not_previously_skipped_dep.NotPreviouslySkippedDep'
+        assert deps[2] == 'airflow.ti_deps.deps.prev_dagrun_dep.PrevDagrunDep'
+        assert deps[3] == 'airflow.ti_deps.deps.ready_to_reschedule.ReadyToRescheduleDep'
+        assert deps[4] == 'airflow.ti_deps.deps.trigger_rule_dep.TriggerRuleDep'
```
Comment (uranusjr) [inline tests/serialization/test_dag_serialization.py]:
```suggestion
        assert deps == [
            'airflow.ti_deps.deps.not_in_retry_period_dep.NotInRetryPeriodDep',
            'airflow.ti_deps.deps.not_previously_skipped_dep.NotPreviouslySkippedDep',
            'airflow.ti_deps.deps.prev_dagrun_dep.PrevDagrunDep',
            'airflow.ti_deps.deps.ready_to_reschedule.ReadyToRescheduleDep',
            'airflow.ti_deps.deps.trigger_rule_dep.TriggerRuleDep',
        ]
```

Same for similar assert statements below.

**Review by uranusjr — CHANGES_REQUESTED** [review]
Logic looks good to me, some coding style comments inline.

Could you also add comments to where `sorted` is called to describe the reason?

**Review by kaxil — CHANGES_REQUESTED** [review]
Which version did you try this with, we had already fixed some of it here:  https://github.com/apache/airflow/pull/15395

**Comment by kaxil** [comment]
And like Ash explains in https://github.com/apache/airflow/pull/15395#discussion_r614418669 , list and tuple's should be sorted.

If there is an issue with how we are storing `deps` we should convert it to a set

**Comment by luoyuliuyin** [comment]
> Which version did you try this with, we had already fixed some of it here: #15395

The version I am using is 2.1.0, and this problem still exists now

**Comment by luoyuliuyin** [comment]
> And like Ash explains in [#15395 (comment)](https://github.com/apache/airflow/pull/15395#discussion_r614418669) , list and tuple's should be sorted.
> 
> If there is an issue with how we are storing `deps` we should convert it to a set

I agree with ashb, `[2, 0, 3]` and` [0, 2, 3]` are not the same.

We now need to serialize dag to json, and then calculate the hash of json_str to determine whether the dag has changed. But json doesn't directly work with sets because the data stored in the set is not stored as properties, we usually convert set to list to solve this problem. So in the current serialization scenario, it is reasonable to convert set to list.

For the case of `sorted(list)` throwing an exception, we need to ensure that the elements in the list are of the Iterable type and that the elements are comparable. In the current modification, both deps and task_group.xxx_ids are str, so they can be compared. At the same time, we are not sure whether the return of `cls._serialize(v)` is comparable, nor whether the semantics will be changed after sorting, so we should not sort in the external function, and we should give the choice to the specific execution function.

File: airflow/serialization/serialized_objects.py (hunk 1 above)
Comment (luoyuliuyin) [reply airflow/serialization/serialized_objects.py]: fixed

File: tests/serialization/test_dag_serialization.py (hunk 2 above)
Comment (luoyuliuyin) [reply tests/serialization/test_dag_serialization.py]: fixed

**Comment by luoyuliuyin** [comment]
> Logic looks good to me, some coding style comments inline.
> 
> Could you also add comments to where `sorted` is called to describe the reason?

fixed

File: airflow/serialization/serialized_objects.py [hunk 3]
```diff
@@ -424,7 +424,10 @@ def serialize_operator(cls, op: BaseOperator) -> Dict[str, Any]:
                     )
 
                 deps.append(f'{module_name}.{klass.__name__}')
-            serialize_op['deps'] = deps
+            # deps needs to be sorted here, because op.deps is a set, which is unstable when traversing,
+            # and the same call may get different results.
+            # When calling json.dumps(self.data, sort_keys=True) to generate dag_hash, misjudgment will occur
+            serialize_op['deps'] = sorted(deps)
```
Comment (kaxil) [inline airflow/serialization/serialized_objects.py:430]: Are you using Custom deps? Can you please update the PR description and title to better reflect the changes now, something along the lines of, Handle Custom deps correct in DAG Serialization or something like that

Comment (luoyuliuyin) [reply airflow/serialization/serialized_objects.py:430]: Thank you for your comments, I have updated the PR description and title.

**Review by uranusjr — APPROVED** [review]

**Comment by kaxil** [comment]
Re-triggering CI

**Comment by luoyuliuyin** [comment]
> Re-triggering CI

Is the CI service  still service not working?

**Comment by potiuk** [comment]
> > Re-triggering CI
> 
> Is the CI service still service not working?

No. The GitHub Packages stability has been addressed yesterday (we switched to a ghcr.io which is much more stable). You need to fix static checks - you have badly formatted files. I recommend installing pre-commit locally (see CONTRIBUTING.rst) then pre-commit will fix it for you automatically when you do `git commit --amend`

**Comment by luoyuliuyin** [comment]
> Re-triggering CI

static checks have passed, and i have updated the PR description and title

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
Good work @luoyuliuyin 👏

**Pull request merged**